### PR TITLE
Self referencing scene bug_fixed

### LINKED ISSUE #175: Fix self-referencing scenes
Scenes are currently able to reference themselves or scenes higher up in the scene graph.
This creates an infinite loop when parsing the tree.

To reproduce:
Create a scene.
Save it.
Instantiate it in it self.
Change scene.
Try to change back.
Crash.

More info curtesy of Ludvig Arlebrink:
You should not be able to instantiate scene to self (scene 1), meaning, the current scene should not come up as an option in the editor, since this will cause an infinite loop. However, if you instantiate another scene (scene 2) and in scene 2 you instantiate scene 1 the same error, infinite loop, will occur.

### PR DESCRIPTION
**Pull request by 1uk45**
fixes #175

### CODE REVIEW

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 6 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![INFO][INFO] [SceneEditor.cpp#L98](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/4cba175d6fca52c3dcff4e8d9a2f1e968bd45ef8/src/Editor/GUI/Editors/SceneEditor.cpp#L98): horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)
1. ![INFO][INFO] [SceneEditor.cpp#L99](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/4cba175d6fca52c3dcff4e8d9a2f1e968bd45ef8/src/Editor/GUI/Editors/SceneEditor.cpp#L99): horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)
1. ![INFO][INFO] [SceneEditor.cpp#L100](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/4cba175d6fca52c3dcff4e8d9a2f1e968bd45ef8/src/Editor/GUI/Editors/SceneEditor.cpp#L100): horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)
1. ![INFO][INFO] [SceneEditor.cpp#L104](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/4cba175d6fca52c3dcff4e8d9a2f1e968bd45ef8/src/Editor/GUI/Editors/SceneEditor.cpp#L104): horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)
1. ![INFO][INFO] [SceneEditor.cpp#L113](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/4cba175d6fca52c3dcff4e8d9a2f1e968bd45ef8/src/Editor/GUI/Editors/SceneEditor.cpp#L113): horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/Engine/Entity/Entity.hpp [hunk 1]
```diff
@@ -59,10 +60,19 @@ class Entity {
         /// Instantiate a scene as a child to this entity.
         /**
          * @param name The name of the scene to instantiate.
+         * @param isSameScene set to true if it is the same scene you want to istantiate.
          * @return The created root entity of the scene.
          */
-        Entity* InstantiateScene(const std::string& name);
+        Entity* InstantiateScene(const std::string& name, const std::string& originScene);
```
Comment (Chainsawkitten) [inline src/Engine/Entity/Entity.hpp]: originScene isn't document

File: src/Engine/Entity/Entity.hpp [hunk 2]
```diff
@@ -59,10 +60,19 @@ class Entity {
         /// Instantiate a scene as a child to this entity.
         /**
          * @param name The name of the scene to instantiate.
+         * @param isSameScene set to true if it is the same scene you want to istantiate.
```
Comment (Chainsawkitten) [inline src/Engine/Entity/Entity.hpp]: isSameScene does not exist

File: src/Engine/Entity/Entity.hpp [hunk 3]
```diff
@@ -59,10 +60,19 @@ class Entity {
...
-        Entity* InstantiateScene(const std::string& name);
+        Entity* InstantiateScene(const std::string& name, const std::string& originScene);
         
+        /// Check if scene already exists in any of json files.
+        /**
+        * @param filename The name of the scene to check.
+        * @param error Set to true inside the function if it allready exists.
+        * @param root The json value of root scene.
+        */
+        void CheckIfSceneExists(std::string filename, bool & error, std::string originScene, Json::Value root);
```
Comment (Chainsawkitten) [inline src/Engine/Entity/Entity.hpp]: originScene is undocumented

File: src/Engine/Entity/Entity.cpp [hunk 4]
```diff
@@ -71,28 +71,70 @@ bool Entity::HasChild(const Entity* check_child, bool deep) const {
...
     if (FileSystem::FileExists(filename.c_str())) {
-        Json::Value root;
-        std::ifstream file(filename);
-        file >> root;
-        file.close();
-        child->Load(root);
-        
-        child->scene = true;
-        child->sceneName = name;
+        root;
```
Comment (Chainsawkitten) [inline src/Engine/Entity/Entity.cpp]: This line does nothing

File: src/Engine/Entity/Entity.cpp [hunk 5]
```diff
@@ -71,28 +71,70 @@ bool Entity::HasChild(const Entity* check_child, bool deep) const {
...
+        std::ifstream file1(filename);
+        file1 >> root;
+
+        CheckIfSceneExists(filename, error, originScene, root);
+
+        if(error == false) {
+            root;
```
Comment (Chainsawkitten) [inline src/Engine/Entity/Entity.cpp]: This line does nothing

File: src/Engine/Entity/Entity.hpp [hunk 6]
```diff
@@ -1,11 +1,12 @@
 #pragma once
-
 #include <map>
 #include <vector>
 #include <typeinfo>
 #include "../Entity/World.hpp"
 #include <json/json.h>
 #include "../Component/SuperComponent.hpp"
+#include <fstream>
+
```
Comment (NinjaDanz3r) [inline src/Engine/Entity/Entity.hpp]: INFO too many consecutive empty lines [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL005)

File: src/Engine/Entity/Entity.cpp [hunk 7]
```diff
@@ -71,28 +71,70 @@ bool Entity::HasChild(const Entity* check_child, bool deep) const {
...
+    if (error)
+    {
+        child->name = "Error loading scene";
+        Log() << "Scene is added in continous loop.";
+    }
+
     return child;
 }
 
+void Entity::CheckIfSceneExists(std::string filename, bool & error, std::string originScene, Json::Value root)
```
Comment (NinjaDanz3r) [inline src/Engine/Entity/Entity.cpp]: MINOR Function parameter 'filename' should be passed by reference. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3ApassedByValue)
MINOR Function parameter 'originScene' should be passed by reference. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3ApassedByValue)

File: src/Engine/Entity/Entity.cpp [hunk 8]
```diff
@@ -71,28 +71,70 @@ bool Entity::HasChild(const Entity* check_child, bool deep) const {
...
-        
-        child->scene = true;
-        child->sceneName = name;
+        root;
+        std::ifstream file1(filename);
+        file1 >> root;
+
+        CheckIfSceneExists(filename, error, originScene, root);
+
+        if(error == false) {
```
Comment (NinjaDanz3r) [inline src/Engine/Entity/Entity.cpp]: INFO keyword 'if' not followed by a single space [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AT008)

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 5 issues
* ![MINOR][MINOR] 3 minor
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 4 issues
* ![MINOR][MINOR] 3 minor
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/Engine/Entity/Entity.cpp (hunk 4 above)
Comment (1uk45) [reply src/Engine/Entity/Entity.cpp]: fixed

File: src/Engine/Entity/Entity.cpp (hunk 5 above)
Comment (1uk45) [reply src/Engine/Entity/Entity.cpp]: fixed

File: src/Engine/Entity/Entity.hpp (hunk 6 above)
Comment (1uk45) [reply src/Engine/Entity/Entity.hpp]: fixed

File: src/Engine/Entity/Entity.cpp (hunk 7 above)
Comment (1uk45) [reply src/Engine/Entity/Entity.cpp]: fixed

Comment (1uk45) [reply src/Engine/Entity/Entity.cpp]: sorry not fixed

Comment (1uk45) [reply src/Engine/Entity/Entity.cpp]: now its fixed

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 6 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![INFO][INFO] [SceneEditor.cpp#L98](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/059e4449cfbe2dccc03bfb1cfda60e27ee696560/src/Editor/GUI/Editors/SceneEditor.cpp#L98): horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)
1. ![INFO][INFO] [SceneEditor.cpp#L99](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/059e4449cfbe2dccc03bfb1cfda60e27ee696560/src/Editor/GUI/Editors/SceneEditor.cpp#L99): horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)
1. ![INFO][INFO] [SceneEditor.cpp#L100](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/059e4449cfbe2dccc03bfb1cfda60e27ee696560/src/Editor/GUI/Editors/SceneEditor.cpp#L100): horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)
1. ![INFO][INFO] [SceneEditor.cpp#L104](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/059e4449cfbe2dccc03bfb1cfda60e27ee696560/src/Editor/GUI/Editors/SceneEditor.cpp#L104): horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)
1. ![INFO][INFO] [SceneEditor.cpp#L113](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/059e4449cfbe2dccc03bfb1cfda60e27ee696560/src/Editor/GUI/Editors/SceneEditor.cpp#L113): horizontal tab used [rule](http://localhost:9000/coding_rules#rule_key=vera%2B%2B%3AL002)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**